### PR TITLE
Added 'lenient ops' module

### PR DESCRIPTION
**Pull request by bluenote10**
@Araq By chance I saw your idea with lenient ops on IRC. I had the same idea before and I was often using stuff like this for numeric-expression-heavy stuff. Would be nice to have this in the stdlib.

I'm only not sure how to name the module. Just `lenient` is maybe not meaningful enough. Alternatives that came to my mind: `auto_ops`, `convert_ops`, `float_int_ops` -- nothing really catchy.

### CODE REVIEW

File: lib/pure/lenientops.nim [hunk 1]
```diff
@@ -0,0 +1,142 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2015 Andreas Rumpf, Dominik Picheta
```
Comment (Araq) [inline lib/pure/lenientops.nim]: Either use your own name here or "Nim contributors"

File: lib/pure/lenientops.nim [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+## type of the float operand. For numerical expressions, the return
+## type is always the type of the float involved in the expresssion,
+## i.e., there is no auto conversion from float32 to float64.
+##
+## Note: In general, auto-converting from int to float loses
+## information, which is why these operators live in a separate
+## module. Use with care.
+
+
+proc `+`[I: SomeInteger, F: SomeReal](i: I, f: F): F = F(i) + f
```
Comment (Araq) [inline lib/pure/lenientops.nim]: These should all be ``inline`` procs, better take no risks. :-)

File: lib/pure/lenientops.nim [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+proc `>=`[I: SomeInteger, F: SomeReal](i: I, f: F): bool {.noSideEffect, inline.} =
+  F(i) >= f
+proc `>=`[I: SomeInteger, F: SomeReal](f: F, i: I): bool {.noSideEffect, inline.} =
+  f >= F(i)
+proc `>`[I: SomeInteger, F: SomeReal](i: I, f: F): bool {.noSideEffect, inline.} =
+  F(i) > f
+proc `>`[I: SomeInteger, F: SomeReal](f: F, i: I): bool {.noSideEffect, inline.} =
+  f > F(i)
+
+when isMainModule:
```
Comment (Araq) [inline lib/pure/lenientops.nim]: Due to the way overloading and scoping works (inner module overloads are preferred in the case of an ambiguity) you need to move these tests into a different module.

Comment (bluenote10) [reply lib/pure/lenientops.nim]: No problem, will do. This will also reveal one of my favorite mistakes: forgetting the `*` ;)

**Comment by bluenote10** [comment]
I have finalized the PR a bit. Things to approve by reviewers:

- I have added warning for `==` operator. This is the only operator which really might cause some true harm, because the use cases of comparing at integer level or on float level are really both relevant. At first I though about removing the operator entirely. But this would look "incomplete" and I think soon people would open issues that `==` is missing here. That's why I decided to make it a warning instead. What do you think?
- I have added a link in the standard library overview, but I wasn't sure if I should place it under "core" or "math" or somewhere else. Where would you prefer it?

**Comment by konqoro** [comment]
Dumb question: Should those use `toFloat` from system.nim?

**Comment by Araq** [comment]
> What do you think?

Remove it and explain in the docs why it is missing.

> Where would you prefer it?

Core.

**Pull request merged**